### ISSUE TITLE
Middleware.firstRun and UniversalBot.beginDialog(address, dialogId)

### ISSUE THREAD

**Issue by almirbi** [Microsoft/BotBuilder#1837]
I am having a [first run middleware](https://github.com/Microsoft/BotBuilder/blob/master/Node/examples/basics-firstRun/app.js) setup in order to pickup and store the address of the user. It is supposed to run only once and not be called ever again for that specific user, as long as the address is valid. Speaking of which, how long is the address of a user valid (```session.message.address```)?

Additionally, when I store the user's address, there is a cron running which checks for certain conditions, and prompts the user with ```UniversalBot.beginDialog```, using the stored address.

The problem is that when _beginDialog_ is called, the dialog stack is reset, and for some reason the session data which holds the version of the first run middleware is also reset (```session.userData[consts.Data.FirstRunVersion]```), so instead of calling the dialog specified in beginDialog, the dialog which is bound to the firstRun middleware is triggered.

What is the recommend workaround for this? Is this the expected behaviour?

```javascript
bot.use(builder.Middleware.firstRun({ version: 1.0, dialogId: '*:/firstRun' }));
bot.dialog('/firstRun', [
  function (session) {
    builder.Prompts.text(session, "Hello, I am saving your address for later use. Say anything to confirm.");
  },
  function (session, results) {
    tmpStorage[session.message.user.name] = { address: session.message.address };
  }
]);

bot.dialog('/', function(session) {
  session.send("Hello there, it's not yet the time. I'll get back to you later.");
});

bot.dialog('/doStuff', function(session) {
  session.send("Hey there again!");
});

setInterval(function() {
  Object.keys(tmpStorage).forEach(function(user) {
    bot.beginDialog(tmpStorage[user].address, '/doStuff', {user});
  });
}, 2 * 1000);
```

This will output "Hello, I am saving your address for later use. Say anything to confirm." every 2 seconds, instead of saying "Hey there again!" every 2 seconds.

Cheers

Almir

**Comment by Stevenic** [comment]
Sorry for the delay. I'm on vacation... A couple of things. The length of time the address is valid really depends on the channel. Some addresses will always be valid while others can periodically change. We've been saying that in general you should probably refresh the address every 15 days or so.

It's expected that calling `beginDialog()` will reset the dialog stack, that's what its supposed to do but it shouldn't be touching userData. I'll have a quick look at the code.

**Comment by Stevenic** [comment]
So looking at the code the only way you would get a new userData object is if the user ID had changed which seems unlikely. I can try to reproduce this but it may be a week or so after I get back from vacation.

**Comment by slidenerd** [comment]
I guess I am the second person facing the same issue

**Comment by aakashvit** [comment]
@Stevenic @carlosscastro I have also similar problem 
https://github.com/Microsoft/BotBuilder/issues/2311

**Comment by Stevenic** [comment]
So one thing I noticed in look at my code and your code. The first run version is only ever saved once the first rin dialog completes. I don't see you calling `endDialog()` from your first run dialog anywhere.

The other thing I'd suggest is looking at the updated [basics-firstRun](https://github.com/Microsoft/BotBuilder/blob/master/Node/examples/basics-firstRun/app.js) example. I've updated it to show a better way of doing first runs that doesn't require the use of middleware. This updated version shouldn't have this issue.
